### PR TITLE
Feat: dependent features

### PR DESCRIPTION
**Pull request by Tymek**
## About the changes

Relates to [roadmap](https://github.com/orgs/Unleash/projects/10) item: https://github.com/Unleash/unleash/issues/2255

### CODE REVIEW

File: lib/src/core/engine.ts [hunk 1]
```diff
@@ -114,12 +114,14 @@ export class ToggleEngine {
...
       }
     );
-
     if (strategyVariant) {
       return strategyVariant;
     }
 
-    if (feature?.variants) {
+    if (
+      (feature?.strategies.length === 0 || hasEnabledStrategy) &&
```
Comment (Tymek) [inline lib/src/core/engine.ts:122]: Fix for another corner case in client spec

File: lib/package.json [hunk 2]
```diff
@@ -55,7 +55,7 @@
     "commander": "11.0.0",
     "murmurhash3js": "3.0.1",
     "semver": "7.5.4",
-    "unleash-client": "4.1.0",
+    "unleash-client": "4.2.0-beta.0",
```
Comment (Tymek) [inline lib/package.json:58]: types-only, so beta here is OK

File: lib/src/utils.d.ts
Comment (Tymek) [inline lib/src/utils.d.ts:1]: Probably commited before by mistake

File: lib/src/evaluateFlags.ts [hunk 3]
```diff
@@ -22,23 +23,63 @@ export const evaluateFlags = (
...
   });
 
+  const toggles = features
+    .filter((feature) => {
+      if (feature.enabled === false) return false;
+
+      if (!feature?.dependencies?.length) return true;
+
+      return feature.dependencies.every((dependency) => {
+        const parentToggle = features.find(
```
Comment (kwasniew) [inline lib/src/evaluateFlags.ts:45]: I'd remove nested ifs
```ts

  return feature.dependencies.every((dependency) => {
    const parentToggle = features.find((toggle) => toggle.name === dependency.feature);

    if (parentToggle?.dependencies?.length) {
      console.warn(`Unleash: Dependency chain issue.`);
      return false;
    }

    if (parentToggle?.enabled && !dependency.enabled) return false;
    if (!parentToggle?.enabled && dependency.enabled) return false;

    if (
      dependency.variants?.length &&
      !dependency.variants.includes(parentToggle?.variant.name)
    ) {
      return false;
    }

    return true;
  });
```

**Review by kwasniew — APPROVED** [review]

**Pull request merged**